mcs/class/corlib/Mono/SafeStringMarshal.cs

```diff
@@ -0,0 +1,46 @@
+//
+// Safe wrapper for a string and its UTF8 encoding
+//
+// Authors:
+//   Aleksey Kliger <aleksey@xamarin.com>
+//   Rodrigo Kumpera <kumpera@xamarin.com>
+//
+// Copyright 2016 Dot net foundation.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Mono  {
+	internal sealed class SafeStringMarshal : IDisposable {
```

Comment:
I don't understand what about the usage makes it better for this to be a struct? If it's a class and you have multiple references and you incorrectly call Dispose multiple times, it's benign.

If it's a struct and you get multiple copies of it, they will share a `marshaled_string` pointer and if you then Dispose the different copies, you'll end up double-freeing in native code.